File: ios/GotchuReviews/GotchuReviews/Models/Invoice.swift
import Foundation

struct VendorTemplate: Codable {
    let vendorNameNormalized: String
    let fieldLabels: [String]
    let sectionOrder: [String]
    let invoiceNumberFormat: String
    let logoPosition: String
    let fontCategory: String
    let dominantColors: [String]
    let tableStyle: String
}

struct ExtractedInvoice: Codable {
    let contractorName: String
    let businessName: String
    let totalAmount: Double
    let currency: String
    let laborCost: Double?
    let materialsCost: Double?
    let hourlyRate: Double?
    let projectDuration: String?
    let invoiceDate: String?
    let description: String
    let lineItems: [LineItem]
    let contractorPhone: String?
    let contractorEmail: String?
    let contractorAddress: String?
    let zipCode: String?
    let vendorTemplate: VendorTemplate?
}

struct LineItem: Codable {
    let description: String
    let amount: Double
    let category: String
}

struct ExtractionResponse: Codable {
    let extracted: ExtractedInvoice
    let documentHash: String
    let vendorFingerprint: String?
    let vendorTemplate: VendorTemplate?
    let contractorMatches: [ContractorMatch]
    let estimatedBreakdown: Bool
}

struct HistoryResponse: Codable {
    let history: [String: [HistoryInvoice]]
    let summary: HistorySummary
}

struct HistoryInvoice: Codable, Identifiable {
    let id: String
    let totalAmount: Double
    let description: String
    let invoiceDate: String?
    let createdAt: String
    let contractor: ContractorBrief
    let review: ReviewBrief?
}

struct ContractorBrief: Codable {
    let id: String
    let name: String
    let businessName: String
    let category: String
    let city: String?
    let state: String?
}

struct ReviewBrief: Codable {
    let id: String
    let overallRating: Double
    let title: String
}

struct HistorySummary: Codable {
    let totalInvoices: Int
    let totalSpent: Double
    let contractorCount: Int
}

struct VendorHistoryResponse: Codable {
    let vendors: [VendorGroup]
    let summary: HistorySummary
}

struct VendorGroup: Codable, Identifiable {
    let contractor: ContractorBrief
    let invoiceCount: Int
    let totalSpent: Double
    let latestInvoiceDate: String
    let invoices: [DetailedInvoice]
    var id: String { contractor.id }
}

struct DetailedInvoice: Codable, Identifiable {
    let id: String
    let documentHash: String
    let totalAmount: Double
    let currency: String
    let laborCost: Double?
    let materialsCost: Double?
    let hourlyRate: Double?
    let projectDuration: String?
    let invoiceDate: String?
    let description: String
    let lineItems: [LineItem]
    let createdAt: String
    let contractor: ContractorBrief
    let review: ReviewBrief?
}

File: android/GotchuReviews/app/src/main/java/com/gotchureviews/app/data/model/Invoice.kt
package com.gotchureviews.app.data.model

import kotlinx.serialization.Serializable

@Serializable
data class VendorTemplate(
    val vendorNameNormalized: String,
    val fieldLabels: List<String>,
    val sectionOrder: List<String>,
    val invoiceNumberFormat: String,
    val logoPosition: String,
    val fontCategory: String,
    val dominantColors: List<String>,
    val tableStyle: String,
)

@Serializable
data class ExtractedInvoice(
    val contractorName: String,
    val businessName: String,
    val totalAmount: Double,
    val currency: String,
    val laborCost: Double? = null,
    val materialsCost: Double? = null,
    val hourlyRate: Double? = null,
    val projectDuration: String? = null,
    val invoiceDate: String? = null,
    val description: String,
    val lineItems: List<LineItem>,
    val contractorPhone: String? = null,
    val contractorEmail: String? = null,
    val contractorAddress: String? = null,
    val zipCode: String? = null,
    val vendorTemplate: VendorTemplate? = null,
)

@Serializable
data class LineItem(
    val description: String,
    val amount: Double,
    val category: String,
    val quantity: Int = 1,
    private val unitPrice: Double? = null,
) {
    val effectiveUnitPrice: Double get() = unitPrice ?: amount
}

@Serializable
data class ExtractionResponse(
    val extracted: ExtractedInvoice,
    val documentHash: String,
    val vendorFingerprint: String? = null,
    val vendorTemplate: VendorTemplate? = null,
    val contractorMatches: List<ContractorMatch>,
    val estimatedBreakdown: Boolean,
)

@Serializable
data class DetailedInvoice(
    val id: String,
    val documentHash: String,
    val totalAmount: Double,
    val currency: String,
    val laborCost: Double? = null,
    val materialsCost: Double? = null,
    val hourlyRate: Double? = null,
    val projectDuration: String? = null,
    val invoiceDate: String? = null,
    val description: String,
    val lineItems: List<LineItem>,
    val createdAt: String,
    val contractor: ContractorBrief,
    val review: ReviewBrief? = null,
)
